src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs

```diff
@@ -0,0 +1,339 @@
+            string tmpKeychainPassphrase = Guid.NewGuid().ToString("N");
+
+            byte[] utf8Passphrase = System.Text.Encoding.UTF8.GetBytes(tmpKeychainPassphrase);
+
+            SafeTemporaryKeychainHandle keychain;
+
+            int osStatus = AppleCryptoNative_SecKeychainCreateTemporary(
+                tmpKeychainPath,
+                utf8Passphrase.Length,
+                utf8Passphrase,
+                out keychain);
+
+            SafeTemporaryKeychainHandle.TrackKeychain(keychain);
+
+            if (osStatus != 0)
+            {
+                keychain.Dispose();
+                throw CreateExceptionForOSStatus(osStatus);
+            }
+
+            return keychain;
+        }
+
+        internal static void SecKeychainDelete(IntPtr handle, bool throwOnError=true)
+        {
+            int osStatus = AppleCryptoNative_SecKeychainDelete(handle);
+
+            if (throwOnError && osStatus != 0)
+            {
+                throw CreateExceptionForOSStatus(osStatus);
+            }
+        }
+    }
+}
+
+namespace System.Security.Cryptography.Apple
+{
+    internal class SafeKeychainItemHandle : SafeHandle
+    {
+        internal SafeKeychainItemHandle()
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            SafeTemporaryKeychainHandle.UntrackItem(handle);
+            Interop.CoreFoundation.CFRelease(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    internal class SafeKeychainHandle : SafeHandle
+    {
+        internal SafeKeychainHandle()
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+        }
+
+        internal SafeKeychainHandle(IntPtr handle)
+            : base(handle, ownsHandle: true)
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.CoreFoundation.CFRelease(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    internal sealed class SafeTemporaryKeychainHandle : SafeKeychainHandle
+    {
+        private static readonly Dictionary<IntPtr, SafeTemporaryKeychainHandle> s_lookup =
+            new Dictionary<IntPtr, SafeTemporaryKeychainHandle>();
+
+        internal SafeTemporaryKeychainHandle()
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            lock (s_lookup)
+            {
+                s_lookup.Remove(handle);
+            }
+
+            Interop.AppleCrypto.SecKeychainDelete(handle, throwOnError: false);
+            return base.ReleaseHandle();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && SafeHandleCache<SafeTemporaryKeychainHandle>.IsCachedInvalidHandle(this))
```

Comment:
Just curious, is the disposing check needed here?  If we're not disposing, we're finalizing, and since this would be a cached handle, wouldn't that necessarily mean it's getting finalized due to the process shutting down such that the static root was dropped?